### PR TITLE
irmin: Trees carry a placeholder proof env

### PR DESCRIPTION
**Pull request by Ngoguey42**
Shipping that feature alone without the rest of the proofs simplifies a lot my git life :-)

It contains half of the modifications on tree.ml caused by the proofs (the other half is the portable tag). 

When an irmin function has trees in both negative and positive position, the positive one(s) should inherit the env of the negative one.

I suspect that the sharing is not perfect at the moment (maybe even bugged), I will investigate and improve this is a later PR, this one strictly a forward port from #1625.

### CODE REVIEW

File: src/irmin/tree.ml [hunk 1]
```diff
@@ -124,6 +124,25 @@ module Make (P : Backend.S) = struct
 
   module Metadata = P.Node.Metadata
 
+  module Env = struct
```
Comment (maiste) [inline src/irmin/tree.ml:127]: Just a small random question: why is it called `Env` and not `Proof`?

File: src/irmin/tree.ml [hunk 2]
```diff
@@ -233,14 +258,14 @@ module Make (P : Backend.S) = struct
 
     let clear t = clear_info t.info
 
-    let of_v (v : v) =
+    let of_v ~env (v : v) =
```
Comment (maiste) [inline src/irmin/tree.ml:261]: Couldn't this be an optional argument with `Env.empty ()` as a default value?

Comment (Ngoguey42) [reply src/irmin/tree.ml:261]: I prefer as it is because if forces us to explicitly acknowledge which `env` is the right one in the rest of the code

Comment (maiste) [reply src/irmin/tree.ml:261]: Fair enough :+1:

**Review by CraigFe — APPROVED** [review]
LGTM, let's merge it.

**Pull request merged**